Adopt dynamicDowncast<> in more editing code

annevk opened the pull request:
#### 00672842afaa5ec635ddeb5c4c7b0fa2ac12df8a
<pre>
Adopt dynamicDowncast&lt;&gt; in more editing code
<a href="https://bugs.webkit.org/show_bug.cgi?id=267498">https://bugs.webkit.org/show_bug.cgi?id=267498</a>

Reviewed by Darin Adler.

This implements a refactoring suggested during review of 272993@main.
Namely to turn the tabSpanNode methods into dynamicDowncast&lt;&gt; wrappers.

It also adopts dynamicDowncast&lt;&gt; in the surrounding code, including in
replaceChildrenWithFragment() which will need careful review.

* Source/WebCore/editing/CompositeEditCommand.cpp:
(WebCore::CompositeEditCommand::isRemovableBlock):
(WebCore::CompositeEditCommand::insertNodeAt):
(WebCore::CompositeEditCommand::replaceSelectedTextInNode):
(WebCore::CompositeEditCommand::positionOutsideTabSpan):
(WebCore::CompositeEditCommand::rebalanceWhitespaceOnTextSubstring):
(WebCore::CompositeEditCommand::deleteInsignificantText):
(WebCore::CompositeEditCommand::moveParagraphs):
* Source/WebCore/editing/Editing.cpp:
(WebCore::tabSpanNode):
(WebCore::parentTabSpanNode):
(WebCore::isTabSpanNode): Deleted.
(WebCore::isTabSpanTextNode): Deleted.
* Source/WebCore/editing/Editing.h:
* Source/WebCore/editing/EditingStyle.cpp:
(WebCore::EditingStyle::init):
(WebCore::EditingStyle::conflictsWithInlineStyleOfElement const):
(WebCore::StyleChange::StyleChange):
* Source/WebCore/editing/InsertTextCommand.cpp:
(WebCore::InsertTextCommand::positionInsideTextNode):
(WebCore::InsertTextCommand::insertTab):
* Source/WebCore/editing/markup.cpp:
(WebCore::StyledMarkupAccumulator::wrapWithNode):
(WebCore::StyledMarkupAccumulator::appendCustomAttributes):
(WebCore::StyledMarkupAccumulator::appendStartTag):
(WebCore::highestAncestorToWrapMarkup):
(WebCore::isPlainTextMarkup):
(WebCore::replaceChildrenWithFragment):

Canonical link: <a href="https://commits.webkit.org/273037@main">https://commits.webkit.org/273037@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/15e36557690c46145539087e9c55bd2b2a6ff87f

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/34028 "1 style error") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/12820 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/35996 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/36663 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/30841 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/35098 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/15205 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/9961 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/29889 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/34528 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/10859 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/30363 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/9458 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/9567 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/30383 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/37971 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/30911 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/30706 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/35670 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/9695 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/7623 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/33572 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/11480 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/7838 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/10260 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/10505 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

annevk commented on Source/WebCore/editing/CompositeEditCommand.cpp:
```diff
@@ -1608,13 +1612,13 @@ bool CompositeEditCommand::breakOutOfEmptyListItem()
     RefPtr<Element> newBlock;
     if (RefPtr blockEnclosingList = listNode->parentNode()) {
-        if (is<HTMLLIElement>(*blockEnclosingList)) { // listNode is inside another list item
+        if (RefPtr liElement = dynamicDowncast<HTMLLIElement>(*blockEnclosingList)) { // listNode is inside another list item
```
The style checker is wrong: it does not consider this conditional to use a brace.

darinadler replied:
Maybe the comment is what confuses it?

annevk replied:
Makes sense. I'll leave that as-is for now.

annevk commented on Source/WebCore/editing/markup.cpp:
```diff
@@ -1481,16 +1478,14 @@ ExceptionOr<void> replaceChildrenWithFragment(ContainerNode& container, Ref<Docu
-
-        Ref protectedContainerChild { *containerChild };
-        return containerNode->replaceChild(fragment, protectedContainerChild);
+        return containerNode->replaceChild(fragment, Ref { *containerChildText });
```
This was quite tricky to get correct and ensure existing tests kept passing. It needs careful review.

darinadler commented on Source/WebCore/editing/EditingStyle.cpp:
```diff
@@ -1784,7 +1784,7 @@ StyleChange::StyleChange(EditingStyle* style, const Position& position)
 
     // Changing the whitespace style in a tab span would collapse the tab into a space.
-    if (isTabSpanTextNode(position.deprecatedNode()) || isTabSpanNode((position.deprecatedNode()))) {
+    if (parentTabSpanNode(position.deprecatedNode()) || tabSpanNode((position.deprecatedNode()))) {
```
There seems to be an extra set of parentheses here.

darinadler commented on Source/WebCore/editing/markup.cpp:
```diff
@@ -1452,9 +1449,9 @@ ExceptionOr<Ref<DocumentFragment>> createContextualFragment(Element& element, co
 }
 
-static inline bool hasOneTextChild(ContainerNode& node)
+static inline RefPtr<Text> dynamicOnlyTextChild(ContainerNode& node)
```
I don’t think “dynamic only text child” expresses “return a single child if it’s text”. I would call it `singleTextChild`, which isn’t perfect either but I think is better.

annevk committed:
Adopt dynamicDowncast<> in more editing code
https://bugs.webkit.org/show_bug.cgi?id=267498

Reviewed by Darin Adler.

This implements a refactoring suggested during review of 272993@main.
Namely to turn the tabSpanNode methods into dynamicDowncast<> wrappers.

It also adopts dynamicDowncast<> in the surrounding code, including in
replaceChildrenWithFragment() which will need careful review.

* Source/WebCore/editing/CompositeEditCommand.cpp:
(WebCore::CompositeEditCommand::isRemovableBlock):
(WebCore::CompositeEditCommand::insertNodeAt):
(WebCore::CompositeEditCommand::replaceSelectedTextInNode):
(WebCore::CompositeEditCommand::positionOutsideTabSpan):
(WebCore::CompositeEditCommand::rebalanceWhitespaceOnTextSubstring):
(WebCore::CompositeEditCommand::deleteInsignificantText):
(WebCore::CompositeEditCommand::moveParagraphs):
* Source/WebCore/editing/Editing.cpp:
(WebCore::tabSpanNode):
(WebCore::parentTabSpanNode):
(WebCore::isTabSpanNode): Deleted.
(WebCore::isTabSpanTextNode): Deleted.
* Source/WebCore/editing/Editing.h:
* Source/WebCore/editing/EditingStyle.cpp:
(WebCore::EditingStyle::init):
(WebCore::EditingStyle::conflictsWithInlineStyleOfElement const):
(WebCore::StyleChange::StyleChange):
* Source/WebCore/editing/InsertTextCommand.cpp:
(WebCore::InsertTextCommand::positionInsideTextNode):
(WebCore::InsertTextCommand::insertTab):
* Source/WebCore/editing/markup.cpp:
(WebCore::StyledMarkupAccumulator::wrapWithNode):
(WebCore::StyledMarkupAccumulator::appendCustomAttributes):
(WebCore::StyledMarkupAccumulator::appendStartTag):
(WebCore::highestAncestorToWrapMarkup):
(WebCore::isPlainTextMarkup):
(WebCore::replaceChildrenWithFragment):

Canonical link: https://commits.webkit.org/273037@main
